ocaml/typing/typecore.ml

```diff
@@ -4016,24 +4022,36 @@ and type_expect_
         if rec_flag = Recursive then
           check_recursive_bindings env pat_exp_list
       in
-      (* If the patterns contain module unpacks, there is a possibility that the
-         types of the let body or let definitions mention types
-         introduced by those unpacks. (The latter can only happen with recursive
-         definitions.) Here, we check for scope escape via both
-         of these pathways (body, definitions).
+      (* The "bound expressions" component of the scope escape check.
 
-         Checking unification within an environment extended with the module
-         bindings allows us to correctly accept more programs. This environment
-         allows unification to identify more cases where a type introduced by
-         the module is equal to a type introduced at an outer scope.
+         This kind of scope escape is relevant only for recursive
+         module definitions.
       *)
+      if rec_flag = Recursive && may_contain_modules then begin
+        List.iter
+          (fun vb ->
+             (* [type_let] already generalized bound expressions' types
+                in-place. We first take an instance before checking scope
+                escape at the outer level to avoid losing generality of
+                types added to [new_env].
+             *)
+             let bound_exp = vb.vb_expr in
+             generalize_structure bound_exp.exp_type;
```

Comment:
Why is this step needed?

Reply:
By "this step", are you referring to just the `generalize_structure`, or to this block of code as a whole?

If you mean the block of code as a whole, I refer to this discussion with garrigue: https://github.com/ocaml/ocaml/pull/12117#discussion_r1154213226. This is the scope escape check for the LHS of recursive first-class module let-bindings. He pointed out some issues with the previous approach.

If you mean just the `generalize_structure`, garrigue suggested that:

> I suppose that the correct thing to do would be to first call `generalize_structure`, to ensure that nodes whose descendant is generalized are themselves generalized, and then call `unify_exp_types` on the result of `Ctype.instance`.

I'm not confident whether this `generalize_structure` is strictly needed.

Reply:
I'm referring just to the `generalize_structure`. I don't see what good it can do. Maybe @lpw25 has a better sense here?

Reply:
`generalize` has been called on all bound expressions' types before this point. (Specifically, on [this line](https://github.com/ocaml/ocaml/blob/trunk/typing/typecore.ml#L5351) or [this line](https://github.com/ocaml/ocaml/blob/trunk/typing/typecore.ml#L5355). Both of these `generalize`s occur at the same level as this `generalize_structure`.) So, I would hazard that it's fine to remove this line, under the assumption that `generalize_structure` after `generalize` is a no-op. What do you think?

If we decide to change this here, I can make the change along with another PR I'm planning to submit upstream to remove it there as well.

It's worth mentioning that this code only runs for recursive first-class module definitions, which are very rare. (I see the argument that we should omit needless calls to `generalize` for clarity, but efficiency does not seem to be at issue.)

Reply:
Yes, I say remove the call. (It's not for efficiency -- it's for clarity and maintainability of code.) If your patch upstream removes this and receives comments saying it's necessary, we'll learn something new.

Reply:
It's now removed and the branch is now rebased. If this is the last unresolved comment, I think we should be good to merge.